src/main/java/com/streamr/client/dataunion/DataUnionClient.java

```diff
@@ -0,0 +1,290 @@
+    }
+
+    protected IERC20 sidechainToken() throws Exception {
+        String tokenAddress = factorySidechain().token().send().getValue();
+        return IERC20.load(tokenAddress, sidechain, sidechainCred, sidechainGasProvider());
+    }
+
+    protected ForeignAMB mainnetAMB(Credentials cred) throws Exception {
+        String amb = factoryMainnet().amb().send().getValue();
+        return ForeignAMB.load(amb, mainnet, cred, mainnetGasProvider());
+    }
+
+    protected ForeignAMB mainnetAMB() throws Exception {
+        return mainnetAMB(mainnetCred);
+    }
+
+    protected HomeAMB sidechainAMB() throws Exception {
+        String amb = factorySidechain().amb().send().getValue();
+        return HomeAMB.load(amb, sidechain, sidechainCred, sidechainGasProvider());
+    }
+
+
+    /**
+     *
+     * @param msgHash
+     * @param sleeptime
+     * @param timeout
+     * @return number of signatures if complete or null if incomplete
+     * @throws Exception
+     */
+    protected BigInteger waitForSidechainAMBAffirmations(final Bytes32 msgHash, long sleeptime, long timeout) throws Exception {
+        Condition affirmationsChange = new Condition(){
+            @Override
+            public Object check() throws Exception {
+                BigInteger requiredSignatures = sidechainAMB().requiredSignatures().send().getValue();
+                BigInteger signatures = sidechainAMB().numMessagesSigned(msgHash).send().getValue();
+                //bit 255 is set in AMB to indicate completion. should the same as signatures >= reqd
+                boolean complete = signatures.testBit(255);
+                signatures = signatures.clearBit(255);
+                if(!complete || signatures.compareTo(requiredSignatures) < 0)
+                    return null;
+                return signatures;
+            }
+        };
+        return (BigInteger) waitForCondition(affirmationsChange, sleeptime, timeout);
+    }
+
+    public BigInteger waitForSidechainBalanceChange(BigInteger initialBalance, String balanceAddress, long sleeptime, long timeout) throws Exception {
+        return waitForErc20BalanceChange(initialBalance, sidechainToken().getContractAddress(), balanceAddress, sidechain, sleeptime, timeout);
+    }
+
+    public BigInteger waitForMainnetBalanceChange(BigInteger initialBalance, String balanceAddress, long sleeptime, long timeout) throws Exception {
+        return waitForErc20BalanceChange(initialBalance, mainnetToken().getContractAddress(), balanceAddress, mainnet, sleeptime, timeout);
+    }
+
+    public Boolean waitForMainnetTx(String txhash, long sleeptime, long timeout) throws Exception {
+        return waitForTx(mainnet, txhash, sleeptime, timeout);
+    }
+
+    public Boolean waitForSidechainTx(String txhash, long sleeptime, long timeout) throws Exception {
+        return waitForTx(sidechain, txhash, sleeptime, timeout);
+    }
+
+
+
+    /*
```

Comment:
Javadoc comments start with `/**` otherwise comment will be ignored by javadoc tool.